Organization of utkscan header files

spaulaus opened the pull request:
These headers were making the directory unreadable and impossible quickly find information. The files have been reorganized and the source files updated to reflect this change.

ksmith0 commented:
This isn't what I expected. I would have gone the other way and had the following:
utkscan
|-analyzers
| |-include
| |-source
|-core
| |-include
| |-source
...

ksmith0 commented:
Also, I dislike putting paths in the include statements as the movement of one header file requires touching everything that uses it. I prefer to let the compiler figure out where the files are.

spaulaus commented:
Fair point (though moot with IDEs.....). I like the suggestion and will make the correction.

spaulaus committed:
Organization of the utkscan sources and headers

This organizational change helps to break the headers and sources into an easier to follow structure. This will also enable some of the utility classes to be moved later.

spaulaus commented:
@ksmith0 How does this strike your fancy?

ksmith0 commented on Scan/utkscan/CMakeLists.txt:
```diff
@@ -34,16 +30,65 @@ if(UTKSCAN_VERBOSE)
+include_directories(include analyzers/include)
+include_directories(include core/include)
+include_directories(include experiment/include)
+include_directories(include processors/include)
```
Why do you repeatedly add the include directory?
Does all the following source need all these includes?

spaulaus replied:
Not all of it, but it's such a mixed bag that it's easier than cherry picking which ones need it and which do not. Remember that we have a mess of circular dependencies here.

ksmith0 replied:
Ok I'll accept that, but you need to drop the repeated addition of the include folder.

ksmith0 reviewed (changes requested):
I would have expected something like

``` cmake
#All utkscan source needs the following headers
include_directories(core/include)

add_subdirectory(core/source)

add_subdirectory(analyzers)
add_subdirectory(processors)
add_subdirectory(experiments)
```

And the CMakeLists.txt in analyzers etc would have

``` cmake
include_directories(include)
add_subdirectory(source)
```

ksmith0 commented on Scan/utkscan/CMakeLists.txt:
```diff
@@ -34,16 +30,65 @@ if(UTKSCAN_VERBOSE)
+#------------------------------------------------------------------------------
+
+#Add the includes from ScanLib to our tree
+include_directories(include ../ScanLib/include)
```
This line shouldn't be here. the ScanLib/include is added in the parent CMakeLists.txt

spaulaus replied:
When removing this line, I get a compilation error 

```
[ 84%] Building CXX object Scan/utkscan/core/source/CMakeFiles/CoreObjects.dir/utkscan.cpp.o
In file included from /home/vincent/programs/sources/utk/paass/Scan/utkscan/core/include/UtkScanInterface.hpp:7:0,
                 from /home/vincent/programs/sources/utk/paass/Scan/utkscan/core/source/utkscan.cpp:4:
/opt/PixieSuite2/include/ScanInterface.hpp:127:24: error: ‘XiaData’ has not been declared
  virtual bool AddEvent(XiaData *event_){ return false; }
                        ^~~~~~~
make[3]: *** [Scan/utkscan/core/source/CMakeFiles/CoreObjects.dir/build.make:447: Scan/utkscan/core/source/CMakeFiles/CoreObjects.dir/utkscan.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/Makefile2:657: Scan/utkscan/core/source/CMakeFiles/CoreObjects.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[ 84%] Built target ProcessorObjects
make[1]: *** [CMakeFiles/Makefile2:559: Scan/utkscan/CMakeFiles/utkscan.dir/rule] Error 2
make: *** [Makefile:266: utkscan] Error 2
```

ksmith0 commented:
I clearly do not know how to use this review process yet. The requested changes are removal of 

``` cmake
include_directories(include ../ScanLib/include)
```

and remove repeated `include` from

``` cmake
include_directories(include analyzers/include)
include_directories(include core/include)
include_directories(include experiment/include)
include_directories(include processors/include)
```

spaulaus committed:
Cleaning up Cmake files

include directories had repeated entries, and we needed to rearrange where the call to ScanLibs include was made.

Commit:
Cleaned up include_directories in Scan.

Set ScanLib/include as an include_direcotry at Scan level. This allows
removal of include_directories directives to ScanLib/include at lower
levels. Removed repeated directives to inlcude_direcotry include.

Commit:
Merge branch 'organize-scan-headers' of https://github.com/pixie16/paass into organize-scan-headers

spaulaus commented:
I think we did exactly the same things....

ksmith0 commented:
I see you resolved the `BUILD_UTKSCAN_TESTS` issue. I propose that before we merge this we (you) move the tests folder to live inside analyzers. Either analyzers/tests or analyzers/source/tests

ksmith0 reviewed (changes requested):
Changes to include_directories is good. Propose we also move tests while we are at it, I imagine the same thing will happen with them in the future.

spaulaus commented:
I can see some merit in this. Though, each of the directories will get their own tests folders eventually. I was trying to centrally locate the tests, but as you say with the includes/sources it may make more sense to colocate all of them together.

ksmith0 commented:
I agree that each one will eventually get a test folder. This is also already true for `Core/tests`. Is the plan to move tests or not?

spaulaus commented:
None of the core files have tests yet. I apologize for not being clear. I agree with the move of tests from utkscan/tests to utkscan/analyzers/tests.

ksmith0 commented:
Yes they [do](https://github.com/pixie16/paass/tree/master/Core/tests), although it is more of a test program. (Some info about use in issue #70)

spaulaus commented:
Oh sorry! We have two "Cores" there's a core for utkscan as well, which is what I assumed was the reference given that this thread is mostly about that!

ksmith0 commented:
Ahh yeah, that is confusing.

spaulaus committed:
Organization of the tests and additions to cmake list files

We have decided to organize the tests to go inside of their specific subgroup folders (e.g. analyzers/tests). This will make distinguishing which test goes with what group a little easier. I have also added new CMakeLists files to the sub folders to aid in expansion of tests in the future. Tests now reside in the tests folder as they will not likely have includes due to their dependence on other classes.

spaulaus committed:
Merge remote-tracking branch 'refs/remotes/origin/organize-scan-headers'

Conflicts:
	Scan/utkscan/CMakeLists.txt

spaulaus commented:
@ksmith0 The request changes have been made. The organization is a little tighter, and more expansible.

ksmith0 reviewed (approved):
Looks good. Squash and merge?

spaulaus commented:
Do it.
